diff --git a/arch/arm/mach-imx/mach-imx7d.c b/arch/arm/mach-imx/mach-imx7d.c
index 3f33597..2ffbb63 100644
--- a/arch/arm/mach-imx/mach-imx7d.c
+++ b/arch/arm/mach-imx/mach-imx7d.c
@@ -61,8 +61,22 @@ static int bcm54220_phy_fixup(struct phy_device *dev)
 	return 0;
 }
 
+static int ksz8081_phy_fixup(struct phy_device *dev)
+{
+	if (dev && dev->interface == PHY_INTERFACE_MODE_MII) {
+		phy_write(dev, 0x1f, 0x8110);
+		phy_write(dev, 0x16, 0x201);
+	} else if (dev && dev->interface == PHY_INTERFACE_MODE_RMII) {
+		phy_write(dev, 0x1f, 0x8190);
+		phy_write(dev, 0x16, 0x202);
+	}
+
+	return 0;
+}
+
 #define PHY_ID_AR8031	0x004dd074
 #define PHY_ID_BCM54220	0x600d8589
+#define PHY_ID_KSZ8081	0x00221560
 
 static void __init imx7d_enet_phy_init(void)
 {
@@ -71,6 +85,8 @@ static void __init imx7d_enet_phy_init(void)
 					   ar8031_phy_fixup);
 		phy_register_fixup_for_uid(PHY_ID_BCM54220, 0xffffffff,
 					   bcm54220_phy_fixup);
+		phy_register_fixup_for_uid(PHY_ID_KSZ8081, 0xfffffffe,
+					   ksz8081_phy_fixup);
 	}
 }
 
@@ -97,8 +113,9 @@ static void __init imx7d_enet_clk_sel(void)
 
 	gpr = syscon_regmap_lookup_by_compatible("fsl,imx7d-iomuxc-gpr");
 	if (!IS_ERR(gpr)) {
+		/* use internal clock generation and output it to PHY */
 		regmap_update_bits(gpr, IOMUXC_GPR1, IMX7D_GPR1_ENET_TX_CLK_SEL_MASK, 0);
-		regmap_update_bits(gpr, IOMUXC_GPR1, IMX7D_GPR1_ENET_CLK_DIR_MASK, 0);
+		regmap_update_bits(gpr, IOMUXC_GPR1, IMX7D_GPR1_ENET_CLK_DIR_MASK, IMX7D_GPR1_ENET2_CLK_DIR_MASK);
 	} else {
 		pr_err("failed to find fsl,imx7d-iomux-gpr regmap\n");
 	}
